signer/gpg2/gpg2.go

```diff
@@ -0,0 +1,241 @@
+	}
+	s.Type = conf.Type
+
+	if conf.ID == "" {
+		return nil, errors.New("gpg2: missing signer ID in signer configuration")
+	}
+	s.ID = conf.ID
+
+	if conf.PrivateKey == "" {
+		return nil, errors.New("gpg2: missing private key in signer configuration")
+	}
+	s.PrivateKey = conf.PrivateKey
+
+	if conf.PublicKey == "" {
+		return nil, errors.New("gpg2: missing public key in signer configuration")
+	}
+	s.PublicKey = conf.PublicKey
+
+	if conf.KeyID == "" {
+		return nil, errors.New("gpg2: missing gpg key ID in signer configuration")
+	}
+	s.KeyID = conf.KeyID
+
+	s.Passphrase = conf.Passphrase
+
+	s.tmpDir, err = createKeyRing(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "gpg2: error creating keyring")
+	}
+	return
+}
+
+// createKeyRing creates a temporary gpg sec and keyrings, loads the
+// private and public keys for the signer, and returns the temporary
+// director holding the rings
+func createKeyRing(s *GPG2Signer) (string, error) {
+	// reuse keyring in tempdir
+	prefix := fmt.Sprintf("autograph_%s_%s", s.Type, s.KeyID)
+
+	dir, err := ioutil.TempDir("", prefix)
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error creating tempdir for keyring")
+	}
+
+	// write the public key to a temp file in our signer's temp dir
+	tmpPublicKeyFile, err := ioutil.TempFile(dir, "gpg2_publickey")
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error creating tempfile for public key")
+	}
+	defer os.Remove(tmpPublicKeyFile.Name())
+	err = ioutil.WriteFile(tmpPublicKeyFile.Name(), []byte(s.PublicKey), 0755)
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error writing public key to tempfile")
+	}
+
+	// write the private key to a temp file in our signer's temp dir
+	tmpPrivateKeyFile, err := ioutil.TempFile(dir, "gpg2_privatekey")
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error creating tempfile for private key")
+	}
+	defer os.Remove(tmpPrivateKeyFile.Name())
+	err = ioutil.WriteFile(tmpPrivateKeyFile.Name(), []byte(s.PrivateKey), 0755)
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error writing private key to tempfile")
+	}
+
+	keyRingPath := filepath.Join(dir, keyRingFilename)
+	secRingPath := filepath.Join(dir, secRingFilename)
+
+	// call gpg to create a new keyring and load the public key in it
+	gpgLoadPublicKey := exec.Command("gpg",
+		"--no-default-keyring",
+		"--keyring", keyRingPath,
+		"--secret-keyring", secRingPath,
+		"--no-tty",
+		"--batch",
+		"--yes",
+		"--import", tmpPublicKeyFile.Name(),
+	)
+	out, err := gpgLoadPublicKey.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrapf(err, "gpg2: failed to load public key into keyring: %s\n%s", err, out)
+	} else {
+		log.Debugf(fmt.Sprintf("gpg2: loaded public key %s", string(out)))
+	}
+
+	// call gpg to load the private key in it
+	gpgLoadPrivateKey := exec.Command("gpg", "--no-default-keyring",
+		"--keyring", keyRingPath,
+		"--secret-keyring", secRingPath,
```

Comment:
I'm concern that we're putting the secring on disk with no way to clean it up should autograph crash and the instance get recycled. Other signers keep their configurations in memory so we have no artifact left on disk that could eventually leak secrets. This is harder to do here, because I don't think Go can register a function to run on shutdown/crash. Maybe the passphrase protects the key sufficiently that we don't have to worry about it? Or maybe we can put the secring on a tmpfs?

Reply:
It is in the tmp dir for the signer.

With Python I'd register [an atexit handler](https://docs.python.org/2/library/atexit.html). With golang we can handle interrupts and clean up on sigterm, sigint, and sigkill though there might be ways to have it not run and I'd expect docker to duplicate the clean up when it's run in container.

Alternatively, the initial implementation https://github.com/mozilla-services/autograph/pull/199/commits/5a4c7ac273cb0e5f6f0dd0dd0d4ad438293f2f69#diff-5c7805a5493d8e612e5a2c5e5974d363L70 creates and cleans up the key and sec rings for each run, so we could revert to that with the caveats that 1) we'd need to namespace and manage rings per request (e.g. by include the request ID in the files) and 2) it would increase per request latency since we're doing more IO for each request. That might be cleanest though.

Reply:
You can't catch ``sigkill``, so that leaves a uncovered corner case.

Reply:
OK added a cleanup handler and checked that the temp dir is cleaned up.

For multiple gpg2 signers or if other signers add a cleanup handler that calls os.Exit we create a race condition. Assuming cleanup is the way we want to go we'll want to:

1) move interrupt handling to main
2) add a cleanup/teardown interface for signers
3) have the main interrupt handler call each that fn for each signer

Reply:
OK went ahead and made the above change so we can support multiple gpg2 signers and confirmed that `kill -SIGKILL` did leave the rings around.

I think our advice should be to run autograph in the container in general and especially if you're testing dep or prod gpg keys, since we've expanded the app boundary to include the filesystem and gpg.